patchwork/scheduling.py

```diff
@@ -220,40 +217,57 @@ class RootQuestionSession(Session):
     # sessions in a real app.
     def __init__(self, scheduler: Scheduler, question: str) -> None:
         super().__init__(scheduler)
-        resulting_context, self.final_answer_promise = scheduler.ask_root_question(question)
-        if resulting_context is None and not self.is_fulfilled():
-            self.current_context = self._choose_next_context()
+        self.current_context, self.final_answer_promise = \
+            scheduler.ask_root_question(question)
+        self.promise_to_advance = self.final_answer_promise
+        self.final_answer = None
+        self.act()
+
+    def choose_promise(self, root: Address) -> Address:
+        """Return unfulfilled promise from hypertext tree with root ``root``.
+
+        Parameters
+        ----------
+        root
+            Address pointing at hypertext or a promise of it.
+        """
+        if not self.sched.db.is_fulfilled(root):
+            return root
+
+        return next((self.choose_promise(child)
+                     for child in self.sched.db.dereference(root).links()),
+                    None)
+
+    # Note: This method is ugly. It's confusing that the action is optional, and
+    # the union return type increases complexity downstream. However, I couldn't
+    # find a better solution and I have to get on with my work. Feel free to
+    # refactor when you feel inspired.
+    def act(self, action: Optional[Action]=None) -> Union[Context, str]:
+        """Take ``action`` in the current context.
+
+        Taking an action entails:
+        1) Execute the action and perhaps advance to the returned context.
+        2) If the final answer is complete, return the final answer.
+        3) If the (new) current context is None, schedule one of the pending
+           contexts.
+
+        If no ``action`` is given, skip the first step.
+        """
+        if action:
+            resulting_context = self.sched.resolve_action(self.current_context,
+                                                          action)
         else:
-            self.current_context = resulting_context
-
-    def _choose_next_context(self):
-        resulting_context = self.sched.choose_context_to_advance_promise(self.final_answer_promise) or \
-                            self.sched.choose_arbitrary_context()
-
-        if resulting_context is None:
-            raise ValueError("Ended up with no work to do but also no answers")
+            resulting_context = self.current_context
 
-        return resulting_context
+        promise_to_advance = self.choose_promise(self.final_answer_promise)
```

Comment:
Maybe it would be clearer if `self.final_answer_promise` was named `self.root_answer_promise`. I will add a commit that renames it. The reason why we need `choose_promise` is that the asker of the root question can't interact with the answer.

If you look at [this scenario](https://github.com/oughtinc/patchwork/pull/7/files/6fdddfaab58d13f6b34a7f0fd8714575bd93f389#r195600518), you will see a `reply $a2` to the root question. I assume that the asker of the root question doesn't want to see this:

```
The final answer is:
 [$1]
```

They can't interact with it, so the code has to check: Are there any locked links/unfulfilled promises in the root answer? If yes, schedule contexts to fulfill them. This is what `self.choose_promise` is for.

Let me know if I should add any clarification that I give you here to the code.

Reply:
Makes sense, thanks! We should probably add a short version of this explanation somewhere in the code, maybe in the docstring for `choose_promise.